[JSC] Implement SuperSampler code in LLInt

Constellation opened the pull request:
#### 5f72014c126b4da2334c8636d7f642e50bdaab6b
<pre>
[JSC] Implement SuperSampler code in LLInt
<a href="https://bugs.webkit.org/show_bug.cgi?id=252159">https://bugs.webkit.org/show_bug.cgi?id=252159</a>
rdar://105382358

Reviewed by Keith Miller and Tadeu Zagallo.

This patch implements op_super_sampler_begin and op_super_sampler_end in LLInt asm instead of function calls from LLInt.
This also adds superSamplerBegin / superSamplerEnd LLInt macros, which can be usable to sample LLInt code.

* Source/JavaScriptCore/bytecode/SuperSampler.h:
* Source/JavaScriptCore/llint/LLIntSlowPaths.cpp:
* Source/JavaScriptCore/llint/LLIntSlowPaths.h:
* Source/JavaScriptCore/llint/LowLevelInterpreter.asm:

Canonical link: <a href="https://commits.webkit.org/260205@main">https://commits.webkit.org/260205@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8f6a84592bf50bac374b3a70870687113dfff9dc

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/107481 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/16543 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/40384 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/116644 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/116065 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/17968 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/7785 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/99645 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/113242 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/13567 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/96748 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/41190 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/95472 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/28374 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/82959 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/96818 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/9549 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/29727 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/96240 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/7548 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/10206 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/6626 "Passed tests") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/46/builds/30808 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/15705 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/49308 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/105088 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/11766 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/26060 "Passed tests") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3826 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/llint/LowLevelInterpreter.asm:
```diff
@@ -406,6 +406,15 @@ macro dispatchOp(size, opcodeName)
 
+macro superSamplerBegin(scratch)
+    leap _g_superSamplerCount, scratch
+    addi 1, [scratch]
```
This is the same way to BaselineJIT / DFGJIT / FTLJIT.

kmiller68 replied:
Do you think we should make it atomic? I guess we’d have to teach the LLInt how to emit those instructions then?

Constellation replied:
Yes, but eventually.

tadeuzagallo reviewed (approved):
r=me

kmiller68 commented on Source/JavaScriptCore/bytecode/SuperSampler.h:
```diff
@@ -31,7 +31,7 @@ namespace JSC {
 class MacroAssembler;
 
-extern JS_EXPORT_PRIVATE std::atomic<uint32_t> g_superSamplerCount;
+extern "C" JS_EXPORT_PRIVATE std::atomic<uint32_t> g_superSamplerCount;
```
I enjoy the idea of extern “C”ing a clearly C++ type. Only good things can be done with that.

kmiller68 reviewed (approved):
r=me too

Constellation committed:
[JSC] Implement SuperSampler code in LLInt
https://bugs.webkit.org/show_bug.cgi?id=252159
rdar://105382358

Reviewed by Keith Miller and Tadeu Zagallo.

This patch implements op_super_sampler_begin and op_super_sampler_end in LLInt asm instead of function calls from LLInt.
This also adds superSamplerBegin / superSamplerEnd LLInt macros, which can be usable to sample LLInt code.

* Source/JavaScriptCore/bytecode/SuperSampler.h:
* Source/JavaScriptCore/llint/LLIntSlowPaths.cpp:
* Source/JavaScriptCore/llint/LLIntSlowPaths.h:
* Source/JavaScriptCore/llint/LowLevelInterpreter.asm:

Canonical link: https://commits.webkit.org/260205@main
